### PR TITLE
added YAML (de)serialization

### PR DESCRIPTION
**Pull request by mizrael**
## Description
- Added missing YAML serializer and deserializer 
- added tests

## Validation

- unit tests

### CODE REVIEW

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Yaml/YamlSerializationFactory.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+using YamlDotNet.Serialization;
+
+namespace Microsoft.PowerPlatform.PowerApps.Persistence.Yaml;
+
+public static class YamlSerializationFactory
+{
+    public static ISerializer CreateSerializer()
+    {
+        var yamlSerializer = new SerializerBuilder()
```
Comment (tehcrashxor) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence/Yaml/YamlSerializationFactory.cs:12]: [.WithNewLine(...)](https://github.com/aaubry/YamlDotNet/blob/master/YamlDotNet/Serialization/SerializerBuilder.cs#L369C9-L373C10) if we want to make sure that we have the same endings on Windows and Linux/Mac

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Models/ControlPropertiesCollection.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.PowerPlatform.PowerApps.Persistence.Models;
+
+public class ControlPropertiesCollection : IReadOnlyDictionary<string, ControlPropertyValue>
```
Comment (tehcrashxor) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence/Models/ControlPropertiesCollection.cs:9]: I'm not sure if there are any benefits to making our own collection type instead of just using `IReadOnlyDictionary<string, ControlPropertyValue>`.

Anything that I'm not seeing which a custom collection would help wtih?

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj [hunk 3]
```diff
@@ -6,4 +6,37 @@
...
+            See https://github.com/microsoft/PowerApps-Tooling/releases for the latest release notes.
+        </PackageReleaseNotes>
+        <Copyright>© Microsoft Corporation. All rights reserved.</Copyright>
+        <GenerateAssemblyInfo>true</GenerateAssemblyInfo>
+        <!-- Workaround for version range https://github.com/NuGet/Home/issues/11842 -->
+        <RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>
+        <NoWarn>$(NoWarn);NU1601</NoWarn>
+    </PropertyGroup>
+
+  <ItemGroup>
```
Comment (tehcrashxor) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj:34]: looks like inconsistent spacing here and below

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Yaml/YamlSerializationFactory.cs (hunk 1 above)
Comment (mizrael) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Yaml/YamlSerializationFactory.cs:12]: added that, it fixed the tests on linux. Thanks!

Comment (tehcrashxor) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Yaml/YamlSerializationFactory.cs:12]: I like the idea of just using `\n` instead of `\r\n` for consistency and better cross-plat support.  Thoughts?

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Models/ControlPropertiesCollection.cs (hunk 2 above)
Comment (mizrael) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Models/ControlPropertiesCollection.cs:9]: @anpetroc suggested it this morning. Honestly, I'm not super opposed to it, but I'm fine with reverting back to `IReadOnlyDictionary` for now

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Yaml/YamlSerializationFactory.cs (hunk 1 above)
Comment (mizrael) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Yaml/YamlSerializationFactory.cs:12]: I'm with you on this. Let's discuss with the team before I make the change

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Models/ControlPropertiesCollection.cs (hunk 2 above)
Comment (tehcrashxor) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Models/ControlPropertiesCollection.cs:9]: Hmm.   I guess it does give a nicer target type to `ControlPropertiesCollectionConverter` than just looking for a dictionary.

Comment (mizrael) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Models/ControlPropertiesCollection.cs:9]: `IEnumerable<KeyValuePair<string, ControlPropertyValue>>` would have worked too in the converter btw

File: src/Microsoft.PowerPlatform.PowerApps.Persistence.Tests/GlobalSuppressions.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+// Licensed under the MIT License.
+
+// This file is used by Code Analysis to maintain SuppressMessage
+// attributes that are applied to this project.
+// Project-level suppressions either have no target or are given
+// a specific target and scoped to a namespace, type, member, etc.
+
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("Naming", "CA1707:Identifiers should not contain underscores", Justification = "Allowed for tests methods")]
```
Comment (petrochuk) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence.Tests/GlobalSuppressions.cs:11]: this is good idea for tests

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Extensions/TypeExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+
+// Copyright (c) Microsoft Corporation.
```
Comment (petrochuk) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence/Extensions/TypeExtensions.cs:5]: Don't need two headers

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj [hunk 6]
```diff
@@ -6,4 +6,39 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <!-- Nuget Properties -->
+  <PropertyGroup>
+    <PackageId>Microsoft.PowerPlatform.PowerApps.Persistence</PackageId>
+    <Authors>Microsoft</Authors>
+    <Company>crmsdk,Microsoft</Company>
+    <Title>PowerPlatform Canvas App Persistence Tools</Title>
```
Comment (petrochuk) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj:14]: Microsoft Power Platform Canvas App Persistence Library?

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj [hunk 7]
```diff
@@ -6,4 +6,39 @@
...
+    <PackageProjectUrl>https://github.com/microsoft/PowerApps-Tooling</PackageProjectUrl>
+    <PackageLicenseExpression>MIT</PackageLicenseExpression>
+    <PackageRequireLicenseAcceptance>true</PackageRequireLicenseAcceptance>
+    <Description>Preview Release</Description>
+    <PackageReleaseNotes>
+      Notice:
+      This package is a preview release - use at your own risk.
+      This package is a .NET Standard 2.0 project, intended to work with .NET Framework 4.7.2 or
```
Comment (petrochuk) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj:22]: looks like copy & paste. we are not doing .NET Standard 2.0

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Extensions/TypeExtensions.cs (hunk 5 above)
Comment (mizrael) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Extensions/TypeExtensions.cs:5]: good catch!

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj [hunk 8]
```diff
@@ -6,4 +6,38 @@
...
+    </PackageReleaseNotes>
+    <Copyright>© Microsoft Corporation. All rights reserved.</Copyright>
+    <GenerateAssemblyInfo>true</GenerateAssemblyInfo>
+    <!-- Workaround for version range https://github.com/NuGet/Home/issues/11842 -->
+    <RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>
+    <NoWarn>$(NoWarn);NU1601</NoWarn>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="YamlDotNet" Version="15.1.0" />
```
Comment (petrochuk) [inline src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj:36]: Let's move versions to root props file. It can be separate PR

**Review by petrochuk — APPROVED** [review]

**Review by tehcrashxor — APPROVED** [review]

File: src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj (hunk 8 above)
Comment (mizrael) [reply src/Microsoft.PowerPlatform.PowerApps.Persistence/Microsoft.PowerPlatform.PowerApps.Persistence.csproj]: I was thinking the same this

**Review by petrochuk — APPROVED** [review]

File: src/Directory.Build.props [hunk 9]
```diff
@@ -1,7 +1,9 @@
 <Project>
+    <Import Project="Versions.props" />
+
     <PropertyGroup>
         <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
-        <TargetFramework>net7.0</TargetFramework>
+        <TargetFramework>$(TargetFrameworkVersion)</TargetFramework>
```
Comment (joem-msft) [inline src/Directory.Build.props:6]: do not use this property. it's recommended to specify the <TargetFrameworks> properties, even if there's only one.
And when there are a small number of projects, don't add yet another property for managing it.

Comment (joem-msft) [reply src/Directory.Build.props:6]: note: TargetFrameworkVersion is one of the properties set by the MSBuild targets, or it looks like one that is.
it's used in pre-sdk-style projects, which this is not.

File: src/Directory.Build.props [hunk 10]
```diff
@@ -1,7 +1,9 @@
 <Project>
+    <Import Project="Versions.props" />
```
Comment (joem-msft) [inline src/Directory.Build.props:2]: for our new projects, please remove uses of Import and shared props like this.
instead, add these to a Directory.Build.props or .targets, which is the modern way.

File: src/Versions.props [hunk 11]
```diff
@@ -0,0 +1,11 @@
+<Project>
+    <PropertyGroup>
```
Comment (joem-msft) [inline src/Versions.props:2]: please lets start using [Central Package Management](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management) pattern for our new projects.

File: src/Versions.props [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+        <TargetFrameworkVersion>net7.0</TargetFrameworkVersion>
```
Comment (joem-msft) [inline src/Versions.props:3]: TargetFrameworkVersion is a predefined msbuild property. Do NOT use it.
It has been replaced by sdk-style projects using <TargetFramework> or <TargetFrameworks>

**Review by joem-msft — CHANGES_REQUESTED** [review]
please apply these comments in order to ensure we're accomplishing our goal of utilizing the latest SDK patterns and practices.
this helps to minimize confusion for opensource project consumers too.
